File: crates/brainwires-stores/src/plan_store.rs
//! Plan Store - Persists execution plans with conversation association
//!
//! Plans are stored via a [`StorageBackend`](brainwires_storage::databases::StorageBackend) for querying and linked to conversations.
//! They can also be exported as Markdown files for human readability.

use anyhow::{Context, Result};
use std::path::PathBuf;
use std::sync::Arc;

use brainwires_core::{PlanMetadata, PlanStatus};
use brainwires_storage::databases::{
    FieldDef, FieldType, FieldValue, Filter, Record, StorageBackend, record_get,
};
use brainwires_storage::embeddings::CachedEmbeddingProvider;

const TABLE_NAME: &str = "plans";

// ── Schema ──────────────────────────────────────────────────────────────

/// Return the backend-agnostic field definitions for the plans table.
pub fn plans_field_defs(embedding_dim: usize) -> Vec<FieldDef> {
    vec![
        FieldDef::required("plan_id", FieldType::Utf8),
        FieldDef::required("conversation_id", FieldType::Utf8),
        FieldDef::required("title", FieldType::Utf8),
        FieldDef::required("task_description", FieldType::Utf8),
        FieldDef::required("plan_content", FieldType::Utf8),
        FieldDef::optional("model_id", FieldType::Utf8),
        FieldDef::required("status", FieldType::Utf8),
        FieldDef::required("executed", FieldType::Boolean),
        FieldDef::required("iterations_used", FieldType::Int32),
        FieldDef::required("created_at", FieldType::Int64),
        FieldDef::required("updated_at", FieldType::Int64),
        FieldDef::optional("file_path", FieldType::Utf8),
        // Branching fields
        FieldDef::optional("parent_plan_id", FieldType::Utf8),
        FieldDef::optional("child_plan_ids", FieldType::Utf8), // JSON array
        FieldDef::optional("branch_name", FieldType::Utf8),
        FieldDef::required("merged", FieldType::Boolean),
        FieldDef::required("depth", FieldType::Int32),
        // Embedding vector
        FieldDef::optional("embedding", FieldType::Vector(embedding_dim)),
    ]
}

/// Arrow schema for the plans table, used by `LanceDatabase` table creation.
pub fn plans_schema() -> std::sync::Arc<arrow_schema::Schema> {
    use arrow_schema::{DataType, Field};

    std::sync::Arc::new(arrow_schema::Schema::new(vec![
        Field::new("plan_id", DataType::Utf8, false),
        Field::new("conversation_id", DataType::Utf8, false),
        Field::new("title", DataType::Utf8, false),
        Field::new("task_description", DataType::Utf8, false),
        Field::new("plan_content", DataType::Utf8, false),
        Field::new("model_id", DataType::Utf8, true),
        Field::new("status", DataType::Utf8, false),
        Field::new("executed", DataType::Boolean, false),
        Field::new("iterations_used", DataType::Int32, false),
        Field::new("created_at", DataType::Int64, false),
        Field::new("updated_at", DataType::Int64, false),
        Field::new("file_path", DataType::Utf8, true),
        // Branching fields
        Field::new("parent_plan_id", DataType::Utf8, true),
        Field::new("child_plan_ids", DataType::Utf8, true),
        Field::new("branch_name", DataType::Utf8, true),
        Field::new("merged", DataType::Boolean, false),
        Field::new("depth", DataType::Int32, false),
    ]))
}

// ── Record conversion helpers ───────────────────────────────────────────

fn to_record(plan: &PlanMetadata) -> Record {
    let child_plan_ids_json = serde_json::to_string(&plan.child_plan_ids).unwrap_or_default();

    vec![
        (
            "plan_id".into(),
            FieldValue::Utf8(Some(plan.plan_id.clone())),
        ),
        (
            "conversation_id".into(),
            FieldValue::Utf8(Some(plan.conversation_id.clone())),
        ),
        ("title".into(), FieldValue::Utf8(Some(plan.title.clone()))),
        (
            "task_description".into(),
            FieldValue::Utf8(Some(plan.task_description.clone())),
        ),
        (
            "plan_content".into(),
            FieldValue::Utf8(Some(plan.plan_content.clone())),
        ),
        ("model_id".into(), FieldValue::Utf8(plan.model_id.clone())),
        (
            "status".into(),
            FieldValue::Utf8(Some(plan.status.to_string())),
        ),
        ("executed".into(), FieldValue::Boolean(Some(plan.executed))),
        (
            "iterations_used".into(),
            FieldValue::Int32(Some(plan.iterations_used as i32)),
        ),
        (
            "created_at".into(),
            FieldValue::Int64(Some(plan.created_at)),
        ),
        (
            "updated_at".into(),
            FieldValue::Int64(Some(plan.updated_at)),
        ),
        ("file_path".into(), FieldValue::Utf8(plan.file_path.clone())),
        (
            "parent_plan_id".into(),
            FieldValue::Utf8(plan.parent_plan_id.clone()),
        ),
        (
            "child_plan_ids".into(),
            FieldValue::Utf8(Some(child_plan_ids_json)),
        ),
        (
            "branch_name".into(),
            FieldValue::Utf8(plan.branch_name.clone()),
        ),
        ("merged".into(), FieldValue::Boolean(Some(plan.merged))),
        ("depth".into(), FieldValue::Int32(Some(plan.depth as i32))),
        (
            "embedding".into(),
            FieldValue::Vector(plan.embedding.clone().unwrap_or_default()),
        ),
    ]
}

fn from_record(r: &Record) -> Result<PlanMetadata> {
    let status_str = record_get(r, "status")
        .and_then(|v| v.as_str())
        .unwrap_or("draft");
    let status = status_str.parse::<PlanStatus>().unwrap_or_default();

    let child_plan_ids: Vec<String> = record_get(r, "child_plan_ids")
        .and_then(|v| v.as_str())
        .and_then(|json| serde_json::from_str(json).ok())
        .unwrap_or_default();

    let embedding = record_get(r, "embedding")
        .and_then(|v| v.as_vector())
        .map(|v| v.to_vec());

    Ok(PlanMetadata {
        plan_id: record_get(r, "plan_id")
            .and_then(|v| v.as_str())
            .context("missing plan_id")?
            .to_string(),
        conversation_id: record_get(r, "conversation_id")
            .and_then(|v| v.as_str())
            .context("missing conversation_id")?
            .to_string(),
        title: record_get(r, "title")
            .and_then(|v| v.as_str())
            .context("missing title")?
            .to_string(),
        task_description: record_get(r, "task_description")
            .and_then(|v| v.as_str())
            .context("missing task_description")?
            .to_string(),
        plan_content: record_get(r, "plan_content")
            .and_then(|v| v.as_str())
            .context("missing plan_content")?
            .to_string(),
        model_id: record_get(r, "model_id")
            .and_then(|v| v.as_str())
            .map(String::from),
        status,
        executed: record_get(r, "executed")
            .and_then(|v| v.as_bool())
            .unwrap_or(false),
        iterations_used: record_get(r, "iterations_used")
            .and_then(|v| v.as_i32())
            .unwrap_or(0) as u32,
        created_at: record_get(r, "created_at")
            .and_then(|v| v.as_i64())
            .context("missing created_at")?,
        updated_at: record_get(r, "updated_at")
            .and_then(|v| v.as_i64())
            .context("missing updated_at")?,
        file_path: record_get(r, "file_path")
            .and_then(|v| v.as_str())
            .map(String::from),
        embedding,
        parent_plan_id: record_get(r, "parent_plan_id")
            .and_then(|v| v.as_str())
            .map(String::from),
        child_plan_ids,
        branch_name: record_get(r, "branch_name")
            .and_then(|v| v.as_str())
            .map(String::from),
        merged: record_get(r, "merged")
            .and_then(|v| v.as_bool())
            .unwrap_or(false),
        depth: record_get(r, "depth").and_then(|v| v.as_i32()).unwrap_or(0) as u32,
    })
}

// ── PlanStore ───────────────────────────────────────────────────────────

/// Store for managing execution plans
pub struct PlanStore<B: StorageBackend = brainwires_storage::databases::lance::LanceDatabase> {
    backend: Arc<B>,
    embeddings: Arc<CachedEmbeddingProvider>,
    /// Directory for plan markdown exports
    plans_dir: Option<PathBuf>,
}

impl<B: StorageBackend> PlanStore<B> {
    /// Create a new plan store
    pub fn new(backend: Arc<B>, embeddings: Arc<CachedEmbeddingProvider>) -> Self {
        Self {
            backend,
            embeddings,
            plans_dir: None,
        }
    }

    /// Create a plan store with a plans directory for markdown exports
    pub fn with_plans_dir(
        backend: Arc<B>,
        embeddings: Arc<CachedEmbeddingProvider>,
        plans_dir: impl Into<PathBuf>,
    ) -> Self {
        Self {
            backend,
            embeddings,
            plans_dir: Some(plans_dir.into()),
        }
    }

    /// Ensure the underlying table exists.
    pub async fn ensure_table(&self) -> Result<()> {
        let dim = self.embeddings.dimension();
        self.backend
            .ensure_table(TABLE_NAME, &plans_field_defs(dim))
            .await
    }

    /// Save a plan (create or update)
    pub async fn save(&self, plan: &PlanMetadata) -> Result<()> {
        // Delete existing plan with same ID (if any)
        let _ = self.delete(&plan.plan_id).await;

        // Generate embedding from the plan content if not already present
        let mut plan = plan.clone();
        if plan.embedding.is_none() {
            let text = format!("{} {}", plan.title, plan.task_description);
            plan.embedding = Some(self.embeddings.embed(&text)?);
        }

        self.backend
            .insert(TABLE_NAME, vec![to_record(&plan)])
            .await
            .context("Failed to save plan")?;

        Ok(())
    }

    /// Get a plan by ID
    pub async fn get(&self, plan_id: &str) -> Result<Option<PlanMetadata>> {
        let filter = Filter::Eq(
            "plan_id".into(),
            FieldValue::Utf8(Some(plan_id.to_string())),
        );
        let records = self
            .backend
            .query(TABLE_NAME, Some(&filter), Some(1))
            .await?;

        match records.first() {
            Some(r) => Ok(Some(from_record(r)?)),
            None => Ok(None),
        }
    }

    /// Get all plans for a conversation
    pub async fn get_by_conversation(&self, conversation_id: &str) -> Result<Vec<PlanMetadata>> {
        let filter = Filter::Eq(
            "conversation_id".into(),
            FieldValue::Utf8(Some(conversation_id.to_string())),
        );
        let records = self.backend.query(TABLE_NAME, Some(&filter), None).await?;

        let mut plans: Vec<PlanMetadata> =
            records.iter().filter_map(|r| from_record(r).ok()).collect();

        // Sort by created_at descending (newest first)
        plans.sort_by(|a, b| b.created_at.cmp(&a.created_at));

        Ok(plans)
    }

    /// List recent plans across all conversations
    pub async fn list_recent(&self, limit: usize) -> Result<Vec<PlanMetadata>> {
        // Fetch more than needed so we can sort and truncate
        let records = self
            .backend
            .query(TABLE_NAME, None, Some(limit * 2))
            .await?;

        let mut plans: Vec<PlanMetadata> =
            records.iter().filter_map(|r| from_record(r).ok()).collect();

        // Sort by created_at descending and take limit
        plans.sort_by(|a, b| b.created_at.cmp(&a.created_at));
        plans.truncate(limit);

        Ok(plans)
    }

    /// Update an existing plan
    pub async fn update(&self, plan: &PlanMetadata) -> Result<()> {
        self.save(plan).await
    }

    /// Delete a plan by ID
    pub async fn delete(&self, plan_id: &str) -> Result<()> {
        let filter = Filter::Eq(
            "plan_id".into(),
            FieldValue::Utf8(Some(plan_id.to_string())),
        );
        self.backend
            .delete(TABLE_NAME, &filter)
            .await
            .context("Failed to delete plan")?;
        Ok(())
    }

    /// Delete all plans for a conversation
    pub async fn delete_by_conversation(&self, conversation_id: &str) -> Result<()> {
        let filter = Filter::Eq(
            "conversation_id".into(),
            FieldValue::Utf8(Some(conversation_id.to_string())),
        );
        self.backend
            .delete(TABLE_NAME, &filter)
            .await
            .context("Failed to delete plans for conversation")?;
        Ok(())
    }

    /// Search plans by semantic similarity to a query string
    pub async fn search(&self, query: &str, limit: usize) -> Result<Vec<PlanMetadata>> {
        let query_embedding = self.embeddings.embed(query)?;

        let scored = self
            .backend
            .vector_search(TABLE_NAME, "embedding", query_embedding, limit, None)
            .await?;

        let plans: Vec<PlanMetadata> = scored
            .iter()
            .filter_map(|sr| from_record(&sr.record).ok())
            .collect();

        Ok(plans)
    }

    /// Export a plan to a markdown file
    ///
    /// Requires `plans_dir` to be set via `with_plans_dir()`.
    /// Returns the path to the created file.
    pub async fn export_to_markdown(&self, plan_id: &str) -> Result<PathBuf> {
        let plans_dir = self.plans_dir.as_ref().ok_or_else(|| {
            anyhow::anyhow!("Plans directory not configured; use with_plans_dir()")
        })?;

        let plan = self
            .get(plan_id)
            .await?
            .ok_or_else(|| anyhow::anyhow!("Plan not found: {}", plan_id))?;

        // Ensure plans directory exists
        std::fs::create_dir_all(plans_dir)?;

        // Get file path
        let file_path = plans_dir.join(format!("{}.md", plan_id));

        // Generate markdown and write to file
        let markdown = plan.to_markdown();
        std::fs::write(&file_path, markdown)
            .with_context(|| format!("Failed to write plan to {}", file_path.display()))?;

        Ok(file_path)
    }

    /// Save a plan and export to markdown in one operation
    pub async fn save_and_export(&self, plan: &mut PlanMetadata) -> Result<PathBuf> {
        let plans_dir = self.plans_dir.as_ref().ok_or_else(|| {
            anyhow::anyhow!("Plans directory not configured; use with_plans_dir()")
        })?;

        // Export to markdown first
        std::fs::create_dir_all(plans_dir)?;
        let file_path = plans_dir.join(format!("{}.md", &plan.plan_id));
        let markdown = plan.to_markdown();
        std::fs::write(&file_path, &markdown)
            .with_context(|| format!("Failed to write plan to {}", file_path.display()))?;

        // Update file_path in plan
        plan.set_file_path(file_path.to_string_lossy().to_string());

        // Save to database
        self.save(plan).await?;

        Ok(file_path)
    }

    /// Load a plan from its markdown file (useful for editing)
    pub fn load_from_markdown(file_path: &std::path::Path) -> Result<String> {
        std::fs::read_to_string(file_path)
            .with_context(|| format!("Failed to read plan from {}", file_path.display()))
    }

    /// Get all child plans (sub-plans/branches) of a plan
    pub async fn get_children(&self, plan_id: &str) -> Result<Vec<PlanMetadata>> {
        let filter = Filter::Eq(
            "parent_plan_id".into(),
            FieldValue::Utf8(Some(plan_id.to_string())),
        );
        let records = self.backend.query(TABLE_NAME, Some(&filter), None).await?;

        let mut plans: Vec<PlanMetadata> =
            records.iter().filter_map(|r| from_record(r).ok()).collect();

        // Sort by created_at ascending
        plans.sort_by(|a, b| a.created_at.cmp(&b.created_at));
        Ok(plans)
    }

    /// Get the full plan hierarchy (parent and all descendants)
    pub async fn get_hierarchy(&self, plan_id: &str) -> Result<Vec<PlanMetadata>> {
        let mut hierarchy = Vec::new();

        // Get the root plan
        if let Some(root) = self.get(plan_id).await? {
            hierarchy.push(root.clone());

            // Recursively get children
            self.collect_descendants(plan_id, &mut hierarchy).await?;
        }

        Ok(hierarchy)
    }

    /// Recursively collect all descendants
    async fn collect_descendants(
        &self,
        plan_id: &str,
        hierarchy: &mut Vec<PlanMetadata>,
    ) -> Result<()> {
        let children = self.get_children(plan_id).await?;
        for child in children {
            let child_id = child.plan_id.clone();
            hierarchy.push(child);
            // Recursively get children of this child (with depth limit)
            if hierarchy.len() < 100 {
                Box::pin(self.collect_descendants(&child_id, hierarchy)).await?;
            }
        }
        Ok(())
    }
}

File: deno/packages/stores/plan_store.ts
/**
 * Plan Store -- persists execution plans with conversation association.
 *
 * Equivalent to Rust's `stores/plan_store.rs` in rullama-storage.
 * @module
 */

import type { PlanMetadata, PlanStatus } from "@rullama/core";
import { parsePlanStatus } from "@rullama/core";
import type { EmbeddingProvider } from "@rullama/core";
import type { StorageBackend } from "@rullama/storage";
import {
  type FieldDef,
  FieldTypes,
  fieldValueAsBool,
  fieldValueAsI32,
  fieldValueAsI64,
  fieldValueAsStr,
  fieldValueAsVector,
  FieldValues,
  Filters,
  optionalField,
  type Record,
  recordGet,
  requiredField,
} from "@rullama/storage";

const TABLE_NAME = "plans";

function plansFieldDefs(embeddingDim: number): FieldDef[] {
  return [
    requiredField("plan_id", FieldTypes.Utf8),
    requiredField("conversation_id", FieldTypes.Utf8),
    requiredField("title", FieldTypes.Utf8),
    requiredField("task_description", FieldTypes.Utf8),
    requiredField("plan_content", FieldTypes.Utf8),
    optionalField("model_id", FieldTypes.Utf8),
    requiredField("status", FieldTypes.Utf8),
    requiredField("executed", FieldTypes.Boolean),
    requiredField("iterations_used", FieldTypes.Int32),
    requiredField("created_at", FieldTypes.Int64),
    requiredField("updated_at", FieldTypes.Int64),
    optionalField("file_path", FieldTypes.Utf8),
    optionalField("parent_plan_id", FieldTypes.Utf8),
    optionalField("child_plan_ids", FieldTypes.Utf8),
    optionalField("branch_name", FieldTypes.Utf8),
    requiredField("merged", FieldTypes.Boolean),
    requiredField("depth", FieldTypes.Int32),
    optionalField("embedding", FieldTypes.Vector(embeddingDim)),
  ];
}

function toRecord(plan: PlanMetadata): Record {
  const childPlanIdsJson = JSON.stringify(plan.child_plan_ids);
  return [
    ["plan_id", FieldValues.Utf8(plan.plan_id)],
    ["conversation_id", FieldValues.Utf8(plan.conversation_id)],
    ["title", FieldValues.Utf8(plan.title)],
    ["task_description", FieldValues.Utf8(plan.task_description)],
    ["plan_content", FieldValues.Utf8(plan.plan_content)],
    ["model_id", FieldValues.Utf8(plan.model_id ?? null)],
    ["status", FieldValues.Utf8(plan.status)],
    ["executed", FieldValues.Boolean(plan.executed)],
    ["iterations_used", FieldValues.Int32(plan.iterations_used)],
    ["created_at", FieldValues.Int64(plan.created_at)],
    ["updated_at", FieldValues.Int64(plan.updated_at)],
    ["file_path", FieldValues.Utf8(plan.file_path ?? null)],
    ["parent_plan_id", FieldValues.Utf8(plan.parent_plan_id ?? null)],
    ["child_plan_ids", FieldValues.Utf8(childPlanIdsJson)],
    ["branch_name", FieldValues.Utf8(plan.branch_name ?? null)],
    ["merged", FieldValues.Boolean(plan.merged)],
    ["depth", FieldValues.Int32(plan.depth)],
    ["embedding", FieldValues.Vector(plan.embedding ?? [])],
  ];
}

function fromRecord(r: Record): PlanMetadata {
  const statusStr = fieldValueAsStr(recordGet(r, "status")!) ?? "draft";
  const status: PlanStatus = parsePlanStatus(statusStr) ?? "draft";

  const childPlanIds: string[] = (() => {
    const json = recordGet(r, "child_plan_ids");
    if (!json) return [];
    const str = fieldValueAsStr(json);
    if (!str) return [];
    try {
      return JSON.parse(str) as string[];
    } catch {
      return [];
    }
  })();

  const embeddingFv = recordGet(r, "embedding");
  const embedding = embeddingFv ? fieldValueAsVector(embeddingFv) : undefined;

  // Build a PlanMetadata-compatible object
  const result = Object.create(null) as PlanMetadata;
  result.plan_id = fieldValueAsStr(recordGet(r, "plan_id")!)!;
  result.conversation_id = fieldValueAsStr(recordGet(r, "conversation_id")!)!;
  result.title = fieldValueAsStr(recordGet(r, "title")!)!;
  result.task_description = fieldValueAsStr(recordGet(r, "task_description")!)!;
  result.plan_content = fieldValueAsStr(recordGet(r, "plan_content")!)!;
  result.model_id = recordGet(r, "model_id")
    ? fieldValueAsStr(recordGet(r, "model_id")!)
    : undefined;
  result.status = status;
  result.executed = fieldValueAsBool(recordGet(r, "executed")!) ?? false;
  result.iterations_used = fieldValueAsI32(recordGet(r, "iterations_used")!) ??
    0;
  result.created_at = fieldValueAsI64(recordGet(r, "created_at")!)!;
  result.updated_at = fieldValueAsI64(recordGet(r, "updated_at")!)!;
  result.file_path = recordGet(r, "file_path")
    ? fieldValueAsStr(recordGet(r, "file_path")!)
    : undefined;
  result.embedding = embedding;
  result.parent_plan_id = recordGet(r, "parent_plan_id")
    ? fieldValueAsStr(recordGet(r, "parent_plan_id")!)
    : undefined;
  result.child_plan_ids = childPlanIds;
  result.branch_name = recordGet(r, "branch_name")
    ? fieldValueAsStr(recordGet(r, "branch_name")!)
    : undefined;
  result.merged = fieldValueAsBool(recordGet(r, "merged")!) ?? false;
  result.depth = fieldValueAsI32(recordGet(r, "depth")!) ?? 0;
  return result;
}

/** Interface for plan store operations. */
export interface PlanStoreI {
  /** Create the backing `plans` table (including its embedding vector column) if it does not already exist. */
  ensureTable(): Promise<void>;
  /** Insert the plan, replacing any stored plan with the same `plan_id`. */
  save(plan: PlanMetadata): Promise<void>;
  /** Look up a plan by `plan_id`; `undefined` when none has that id. */
  get(planId: string): Promise<PlanMetadata | undefined>;
  /** Every plan whose `conversation_id` matches, newest `created_at` first. */
  getByConversation(conversationId: string): Promise<PlanMetadata[]>;
  /** The `limit` most recently created plans, newest first. */
  listRecent(limit: number): Promise<PlanMetadata[]>;
  /** Delete one plan by `plan_id` (no-op when absent). */
  delete(planId: string): Promise<void>;
  /** Delete every plan belonging to a conversation. */
  deleteByConversation(conversationId: string): Promise<void>;
}

/** Store for managing execution plans. */
export class PlanStore implements PlanStoreI {
  /**
   * Create a store over `backend`; call {@link ensureTable} before use.
   * @param backend Storage backend that holds the `plans` table.
   * @param embeddings Provider used to embed plans for {@link search}; its
   *   `dimension` fixes the width of the table's `embedding` column.
   */
  constructor(
    private readonly backend: StorageBackend,
    private readonly embeddings: EmbeddingProvider,
  ) {}

  /** Create the `plans` table on the backend, sized to the embedding provider's `dimension`. */
  async ensureTable(): Promise<void> {
    await this.backend.ensureTable(
      TABLE_NAME,
      plansFieldDefs(this.embeddings.dimension),
    );
  }

  /**
   * Delete any stored row with the same `plan_id`, then insert `plan`. When
   * `plan.embedding` is unset it is computed from `title` + `task_description`
   * first and written back onto the passed object.
   */
  async save(plan: PlanMetadata): Promise<void> {
    try {
      await this.delete(plan.plan_id);
    } catch { /* ignore */ }

    // Generate embedding if not already present
    if (!plan.embedding) {
      const text = `${plan.title} ${plan.task_description}`;
      plan.embedding = await this.embeddings.embed(text);
    }

    await this.backend.insert(TABLE_NAME, [toRecord(plan)]);
  }

  /** Query the backend for the single record whose `plan_id` matches. */
  async get(planId: string): Promise<PlanMetadata | undefined> {
    const filter = Filters.Eq("plan_id", FieldValues.Utf8(planId));
    const records = await this.backend.query(TABLE_NAME, filter, 1);
    return records.length > 0 ? fromRecord(records[0]) : undefined;
  }

  /** Query every record whose `conversation_id` matches and sort by `created_at` descending. */
  async getByConversation(conversationId: string): Promise<PlanMetadata[]> {
    const filter = Filters.Eq(
      "conversation_id",
      FieldValues.Utf8(conversationId),
    );
    const records = await this.backend.query(TABLE_NAME, filter);
    const plans = records.map(fromRecord);
    plans.sort((a, b) => b.created_at - a.created_at);
    return plans;
  }

  /**
   * Fetch up to `2 * limit` rows from the backend (unordered), sort them by
   * `created_at` descending, and return the first `limit`.
   */
  async listRecent(limit: number): Promise<PlanMetadata[]> {
    const records = await this.backend.query(TABLE_NAME, undefined, limit * 2);
    const plans = records.map(fromRecord);
    plans.sort((a, b) => b.created_at - a.created_at);
    return plans.slice(0, limit);
  }

  /** Delete the record whose `plan_id` matches. */
  async delete(planId: string): Promise<void> {
    const filter = Filters.Eq("plan_id", FieldValues.Utf8(planId));
    await this.backend.delete(TABLE_NAME, filter);
  }

  /** Delete every record whose `conversation_id` matches. */
  async deleteByConversation(conversationId: string): Promise<void> {
    const filter = Filters.Eq(
      "conversation_id",
      FieldValues.Utf8(conversationId),
    );
    await this.backend.delete(TABLE_NAME, filter);
  }

  /**
   * Search plans by semantic similarity: embeds `query` and runs a vector
   * search on the `embedding` column.
   * @param limit Maximum number of plans to return.
   * @returns Matching plans in the backend's ranking order (scores are not returned).
   */
  async search(query: string, limit: number): Promise<PlanMetadata[]> {
    const queryEmbedding = await this.embeddings.embed(query);
    const scored = await this.backend.vectorSearch(
      TABLE_NAME,
      "embedding",
      queryEmbedding,
      limit,
    );
    return scored.map((sr) => fromRecord(sr.record));
  }
}

/** In-memory plan store for testing. */
export class InMemoryPlanStore implements PlanStoreI {
  private plans: Map<string, PlanMetadata> = new Map();

  /** No-op: there is no table to create in memory. */
  async ensureTable(): Promise<void> {
    await Promise.resolve();
  }

  /** Store `plan` by reference (no copy) under its `plan_id`, replacing any existing entry. */
  async save(plan: PlanMetadata): Promise<void> {
    this.plans.set(plan.plan_id, plan);
    await Promise.resolve();
  }

  /** Return the stored plan for `planId`, if any. */
  async get(planId: string): Promise<PlanMetadata | undefined> {
    return await Promise.resolve(this.plans.get(planId));
  }

  /** Stored plans whose `conversation_id` matches, sorted by `created_at` descending. */
  async getByConversation(conversationId: string): Promise<PlanMetadata[]> {
    const plans = [...this.plans.values()]
      .filter((p) => p.conversation_id === conversationId);
    plans.sort((a, b) => b.created_at - a.created_at);
    return await Promise.resolve(plans);
  }

  /** All stored plans sorted by `created_at` descending, truncated to `limit`. */
  async listRecent(limit: number): Promise<PlanMetadata[]> {
    const plans = [...this.plans.values()];
    plans.sort((a, b) => b.created_at - a.created_at);
    return await Promise.resolve(plans.slice(0, limit));
  }

  /** Remove the plan from the map (no-op when absent). */
  async delete(planId: string): Promise<void> {
    this.plans.delete(planId);
    await Promise.resolve();
  }

  /** Remove every stored plan whose `conversation_id` matches. */
  async deleteByConversation(conversationId: string): Promise<void> {
    for (const [id, p] of this.plans) {
      if (p.conversation_id === conversationId) this.plans.delete(id);
    }
    await Promise.resolve();
  }
}
